# Chapter 2: Variables, Expressions, and Statements
**Christopher Hsiao - v1.0**

**variables** are the tools used to store, manipulate, and return data. Using a variables name refers to its value.

## 2.1 Assignment

**Assignment** creates a new variable and *assigns* it a value. Assignment is done using the `=` operator.

Let's try assigning some values.
    
    >>> message = 'Hello, OpenDNS!'
    Hello, OpenDNS!
    >>> num = 17
    17

This allows us to store values for later use. Now, instead of printing the string `'Hello, OpenDNS'`, we can print the variable `message`, and it will give us its value.

    >>> print(message)
    Hello, OpenDNS!
    >>> print(num)
    17

---

## 2.2 Variable Names

Learning what to call your variables is actually a very important skill that, for the most part, is picked up through experience. However, there are some rules that must be followed just to first make sure your variable name is valid.

**What's Legal:** 
* Variable can be any length. 
* They can't *begin* with a number, but can have both numbers and letters within them. 
* You can also use both lower case and upper case letters in your variable names. 
* You'll often see the underscore `_` used, since variable names can't have spaces.

Invalid variable names return a `SyntaxError`.

**Beginning with a number:**

In [10]:
%%python3
1variableName = 3

  File "<stdin>", line 1
    1variableName = 3
                ^
SyntaxError: invalid syntax


**Invalid symbols:**

In [8]:
%%python3
hello@ = 5

  File "<stdin>", line 1
    hello@ = 5
           ^
SyntaxError: invalid syntax


Another condition for an invalid variable name is using a *keyword*. Keywords are words that have special predefined meanings and uses within `Python`. As a matter of fact, all high level languages have *keywords* of some form. If you attempt to assign a keyword a value as if it were a variable, you'll also get a `SyntaxError`.

In [9]:
%%python3
class = 'Physics 101'

  File "<stdin>", line 1
    class = 'Physics 101'
          ^
SyntaxError: invalid syntax


Here's a list of all the keywords in `Python 3`:

    False  class    finally is       return
    None   continue for     lambda   try
    True   def      from    nonlocal while
    and    del      global  not      with
    as     elif     if      or       yield
    assert else     import  pass
    break  except   in      raise
    
Don't worry, you won't need to memorize these. Most text editors or development environments are highlighted, underlined, or displayed in a different color.

---

## 2.3 Expressions and Statements 

An **expression** is a combination of values, variables, and operators. Some legal expressions are:

In [13]:
42

42

In [19]:
n = 17
n + 25

42

A **statement** is a bit of code that has some effect, such as initializing a variable, or printing some value.

In [20]:
x = 17
print(x)

17


Here, we first assign the value 17 to the variable `x`, then call the print() function on it to print the value of `x`.

---

## 2.4 Script Mode

Thus far, the commands we've run were all done in **interactive mode** through the `interpreter`. While this is useful for quickly executing small snippets of code, it can be very cumbersome when attempting to execute anything more than a few lines.

Instead, what is most commonly done is to have the code saved to a file known as a **script**, which typically takes the form of a file ending in **.py**. 

Expressions in the interpreter have the effect of outputting the value of the expression. For example, if you were using the `Python 3` interpreter as a calculator to calculate `25 * 49`, you'd see something like this:

In [22]:
25*49

1225

However, if you have a file with an expression in it, there won't be a valued displayed. For example, we have a file called `expressionDemo.py`, which stores the above single line of code (`25 * 49`). We can execute scripts with 

    python3 [script.py]
    
So let's give it a try. First let's verify the code inside the file with `cat`.

In [31]:
%%bash
cat expressionDemo.py

25*49


Great, now let's actually run it.

In [30]:
%%bash
python3 expressionDemo.py

<div class="alert alert-info">Recall we need to execute the script from the <code>bash</code> using the <code>python3</code> command.</div> 

Note how there is no output. If we want output from expressions, we need to explicitely add `print()` statements to our code. 

If we modify the code in `expressionDemo.py` to `print(25 * 49)`, then we will see the output of `1225`.

### Exercise: Write Your First Script!

